core/app/models/concerns/spree/user_methods.rb

```diff
@@ -90,6 +90,21 @@ def can_be_deleted?
       orders.none?
     end
 
+    # Updates the roles in keeping with the given ability's permissions
+    #
+    # Roles that are not accessible to the given ability will be ignored. It
+    # also ensure not to remove non accessible roles when assigning new
+    # accessible ones.
+    #
+    # @param given_roles [Spree::Role]
+    # @param ability [Spree::Ability]
+    def update_spree_roles(given_roles, ability:)
+      accessible_roles = Spree::Role.accessible_by(ability)
```

Comment:
I agree it's sub-optimal. However, the controller is neither a good place for it. I checked, and it looks like we're not allowing role edition through the API, but if at some point we added it, we'd want this logic to be reusable, and in the absence of any other layer, only the model can cover our backs.

I also realized that when working on it, and I thought it might be a good use case for the spike on the service layer. What do you think?

Reply:
Yeah, on one side we have non-trivial code with the accessible roles "math", on the other a leak of responsibilities, a service helper class would be great, something that could potentially be used in both controllers. 

Alternatively a generic helper for handling collection updates could also do the trick (unless this is the one and only case in which we need this).

Reply:
If that's ok with you, let's keep it for now in the model. In the end, both the receiver and all the arguments are model instances. Let's where the spike about service objects leads us.